app/upgrades/v15/upgrade_test.go

```diff
@@ -67,6 +78,73 @@ func (suite *UpgradeTestSuite) TestMigrateNextPoolIdAndCreatePool() {
 	suite.Require().Equal(gammPoolCreationFee, poolmanagerPoolCreationFee)
 }
 
+
+func (suite *UpgradeTestSuite) TestMigrateBalancerToStablePools() {
+	suite.SetupTest() // reset
+
+	ctx := suite.Ctx
+	gammKeeper := suite.App.GAMMKeeper
+	poolmanagerKeeper := suite.App.PoolManagerKeeper
+	// bankKeeper := suite.App.BankKeeper
+	testAccount := suite.TestAccs[0]
+
+	// Mint some assets to the accounts.
+	suite.FundAcc(testAccount, DefaultAcctFunds)
+
+	// Create the balancer pool
+	swapFee, err := sdk.NewDecFromStr("0.003")
+	exitFee, err := sdk.NewDecFromStr("0.025")
+	poolID, err := suite.App.PoolManagerKeeper.CreatePool(
+		suite.Ctx,
+		balancer.NewMsgCreateBalancerPool(suite.TestAccs[0],
+		balancer.PoolParams{
+			SwapFee: swapFee,
+			ExitFee: exitFee,
+		},
+		[]balancertypes.PoolAsset{
+			{
+				Weight: sdk.NewInt(100),
+				Token:  sdk.NewCoin("foo", sdk.NewInt(5000000)),
+			},
+			{
+				Weight: sdk.NewInt(200),
+				Token:  sdk.NewCoin("bar", sdk.NewInt(5000000)),
+			},
+		},
+		""),
+	)
+	_ = poolID
+	suite.Require().NoError(err)
+
+	balancerPool, err := gammKeeper.GetPool(suite.Ctx, poolID)
+	suite.Require().NoError(err)
+	balancerShares := balancerPool.GetTotalShares()
+
+	// join pool
+	sharesRequested := types.OneShare.MulRaw(1)
+	tokenInMaxs := sdk.Coins{}
+	_, _, err = gammKeeper.JoinPoolNoSwap(suite.Ctx, testAccount, poolID, sharesRequested, tokenInMaxs)
+	suite.Require().Equal(sharesRequested.String(), bankKeeper.GetBalance(suite.Ctx, testAccount, "gamm/pool/1").Amount.String())
+	liquidity := gammKeeper.GetTotalLiquidity(suite.Ctx)
+	suite.Require().Equal("15000bar,15000foo", liquidity.String())
+
+	// test migrating the balancer pool to a stable pool
+	v15.MigrateBalancerPoolToSolidlyStable(ctx, gammKeeper, poolmanagerKeeper, poolID)
+
+	// check that the pool is now a stable pool
+	stablepool, err := gammKeeper.GetPool(ctx, poolID)
+	suite.Require().NoError(err)
+	suite.Require().Equal(stablepool.GetType(), poolmanagertypes.Stableswap)
+
+	// check that a user can withdraw from the pool
```

Comment:
any pointers on functions I should use to test here?

Reply:
I think we should perform these checks on the e2e level.

1. Add the pools with your IDs to the genesis here: https://github.com/osmosis-labs/osmosis/blob/cbc5a44f932a36c11b156e27f2d1c122ef1b9ed3/tests/e2e/initialization/config.go#L368-L392
2. LP to all of the migrated pools in this function: https://github.com/osmosis-labs/osmosis/blob/cbc5a44f932a36c11b156e27f2d1c122ef1b9ed3/tests/e2e/configurer/upgrade.go#L108
3. Write a test in `e2e_test.go` doing swaps and exits from the migrated pools. Similar to: https://github.com/osmosis-labs/osmosis/blob/cbc5a44f932a36c11b156e27f2d1c122ef1b9ed3/tests/e2e/e2e_test.go#L330-L355

I can help you with step 1 tomorrow as it will also require additional work related to e2e setup

Reply:
Actually, we will have to do this anyways as our e2e suite wouldn't pass without the pools with the given ids being setup for upgrade testing.

e2e test is failing with the following error right now:
```
[90m3:15AM[0m [32mINF[0m starting ABCI with Tendermint
[90m3:15AM[0m [32mINF[0m Upgrading IAVL storage for faster queries + execution on live state. This may take a while [36mcommit=[0m436F6D6D697449447B5B5D3A307D [36mis_lazy=[0mfalse [36mstore_key=[0m"KVStoreKey{0xc000f309a0, concentratedliquidity}" [36mversion=[0m118
[90m3:15AM[0m [32mINF[0m Upgrading IAVL storage for faster queries + execution on live state. This may take a while [36mcommit=[0m436F6D6D697449447B5B5D3A307D [36mis_lazy=[0mfalse [36mstore_key=[0m"KVStoreKey{0xc000f30b60, interchainquery}" [36mversion=[0m118
[90m3:15AM[0m [32mINF[0m Upgrading IAVL storage for faster queries + execution on live state. This may take a while [36mcommit=[0m436F6D6D697449447B5B5D3A307D [36mis_lazy=[0mfalse [36mstore_key=[0m"KVStoreKey{0xc000f30b70, packetfowardmiddleware}" [36mversion=[0m118
[90m3:15AM[0m [32mINF[0m Upgrading IAVL storage for faster queries + execution on live state. This may take a while [36mcommit=[0m436F6D6D697449447B5B5D3A307D [36mis_lazy=[0mfalse [36mstore_key=[0m"KVStoreKey{0xc000f309b0, poolmanager}" [36mversion=[0m118
[90m3:15AM[0m [32mINF[0m Upgrading IAVL storage for faster queries + execution on live state. This may take a while [36mcommit=[0m436F6D6D697449447B5B5D3A307D [36mis_lazy=[0mfalse [36mstore_key=[0m"KVStoreKey{0xc000f30b40, protorev}" [36mversion=[0m118
[90m3:15AM[0m [32mINF[0m Upgrading IAVL storage for faster queries + execution on live state. This may take a while [36mcommit=[0m436F6D6D697449447B5B5D3A307D [36mis_lazy=[0mfalse [36mstore_key=[0m"KVStoreKey{0xc000f30b30, valsetpref}" [36mversion=[0m118
[90m3:15AM[0m [32mINF[0m service start [36mimpl=[0mmultiAppConn [36mmodule=[0mproxy [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m service start [36mconnection=[0mquery [36mimpl=[0mlocalClient [36mmodule=[0mabci-client [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m service start [36mconnection=[0msnapshot [36mimpl=[0mlocalClient [36mmodule=[0mabci-client [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m service start [36mconnection=[0mmempool [36mimpl=[0mlocalClient [36mmodule=[0mabci-client [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m service start [36mconnection=[0mconsensus [36mimpl=[0mlocalClient [36mmodule=[0mabci-client [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m service start [36mimpl=[0mEventBus [36mmodule=[0mevents [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m service start [36mimpl=[0mPubSub [36mmodule=[0mpubsub [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m service start [36mimpl=[0mIndexerService [36mmodule=[0mtxindex [36mmsg=[0m{}
[90m3:15AM[0m [32mINF[0m ABCI Handshake App Info [36mhash=[0m"\x0e%6   \x1b W\t}O4D \" \x01\x02  +n 3\aC*5tv " [36mheight=[0m117 [36mmodule=[0mconsensus [36mprotocol-version=[0m0 [36msoftware-version=[0m
[90m3:15AM[0m [32mINF[0m ABCI Replay Blocks [36mappHeight=[0m117 [36mmodule=[0mconsensus [36mstateHeight=[0m117 [36mstoreHeight=[0m118
[90m3:15AM[0m [32mINF[0m Replay last block using real app [36mmodule=[0mconsensus
[90m3:15AM[0m [32mINF[0m applying upgrade "v15" at height: 118
panic: pool with ID 833 does not exist

goroutine 1 [running]:
github.com/osmosis-labs/osmosis/v14/app/upgrades/v15.migrateBalancerPoolToSolidlyStable({{0x2f65168, 0xc000120000}, {0x2f7c300, 0xc00109bb80}, {{0xb, 0x0}, {0xc00126dd30, 0xb}, 0x76, {0x26fbde10, ...}, ...}, ...}, ...)
	github.com/osmosis-labs/osmosis/v14/app/upgrades/v15/upgrades.go:81 +0x545
github.com/osmosis-labs/osmosis/v14/app/upgrades/v15.migrateBalancerPoolsToSolidlyStable(...)
	github.com/osmosis-labs/osmosis/v14/app/upgrades/v15/upgrades.go:73
github.com/osmosis-labs/osmosis/v14/app/upgrades/v15.CreateUpgradeHandler.func1({{0x2f65168, 0xc000120000}, {0x2f7c300, 0xc00109bb80}, {{0xb, 0x0}, {0xc00126dd30, 0xb}, 0x76, {0x26fbde10, ...}, ...}, ...}, ...)
	github.com/osmosis-labs/osmosis/v14/app/upgrades/v15/upgrades.go:55 +0x55b
github.com/cosmos/cosmos-sdk/x/upgrade/keeper.Keeper.ApplyUpgrade({{0xc000d39608, 0x12}, 0xc001436f30, {0x2f48c58, 0xc000f307f0}, {0x2f79bc0, 0xc000fc3d60}, 0xc001437770, {0x2f48b90, 0xc001251520}}, ...)
	github.com/cosmos/cosmos-sdk@v0.46.10/x/upgrade/keeper/keeper.go:286 +0x19e
github.com/cosmos/cosmos-sdk/x/upgrade.BeginBlocker({{0xc000d39608, 0x12}, 0xc001436f30, {0x2f48c58, 0xc000f307f0}, {0x2f79bc0, 0xc000fc3d60}, 0xc001437770, {0x2f48b90, 0xc001251520}}, ...)
	github.com/cosmos/cosmos-sdk@v0.46.10/x/upgrade/abci.go:59 +0x5c5
github.com/cosmos/cosmos-sdk/x/upgrade.AppModule.BeginBlock(...)
	github.com/cosmos/cosmos-sdk@v0.46.10/x/upgrade/module.go:130
github.com/cosmos/cosmos-sdk/types/module.(*Manager).BeginBlock(_, {{0x2f65168, 0xc000120000}, {0x2f7c300, 0xc00109bb80}, {{0xb, 0x0}, {0xc00126dd30, 0xb}, 0x76, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.46.10/types/module/module.go:500 +0x378
github.com/osmosis-labs/osmosis/v14/app.(*OsmosisApp).BeginBlocker(_, {{0x2f65168, 0xc000120000}, {0x2f7c300, 0xc00109bb80}, {{0xb, 0x0}, {0xc00126dd30, 0xb}, 0x76, ...}, ...}, ...)
	github.com/osmosis-labs/osmosis/v14/app/app.go:346 +0xbb
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).BeginBlock(_, {{0xc001082760, 0x20, 0x20}, {{0xb, 0x0}, {0xc00126dd30, 0xb}, 0x76, {0x26fbde10, ...}, ...}, ...})
	github.com/cosmos/cosmos-sdk@v0.46.10/baseapp/abci.go:206 +0x8c3
github.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(_, {{0xc001082760, 0x20, 0x20}, {{0xb, 0x0}, {0xc00126dd30, 0xb}, 0x76, {0x26fbde10, ...}, ...}, ...})
	github.com/tendermint/tendermint@v0.34.24/abci/client/local_client.go:280 +0x118
github.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(_, {{0xc001082760, 0x20, 0x20}, {{0xb, 0x0}, {0xc00126dd30, 0xb}, 0x76, {0x26fbde10, ...}, ...}, ...})
	github.com/tendermint/tendermint@v0.34.24/proxy/app_conn.go:81 +0x55
github.com/tendermint/tendermint/state.execBlockOnProxyApp({0x2f666a0?, 0xc000c3f3e0}, {0x2f71610, 0xc000c41040}, 0xc0000261e0, {0x2f7b470, 0xc000eba030}, 0x75?)
	github.com/tendermint/tendermint@v0.34.24/state/execution.go:307 +0x3dd
github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(_, {{{0xb, 0x0}, {0xc001185e50, 0x7}}, {0xc001185e60, 0xb}, 0x1, 0x75, {{0xc000fec180, ...}, ...}, ...}, ...)
	github.com/tendermint/tendermint@v0.34.24/state/execution.go:140 +0x171
github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlock(_, {{{0xb, 0x0}, {0xc001185e50, 0x7}}, {0xc001185e60, 0xb}, 0x1, 0x75, {{0xc000fec180, ...}, ...}, ...}, ...)
	github.com/tendermint/tendermint@v0.34.24/consensus/replay.go:503 +0x23c
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(_, {{{0xb, 0x0}, {0xc001185e50, 0x7}}, {0xc001185e60, 0xb}, 0x1, 0x75, {{0xc000fec180, ...}, ...}, ...}, ...)
	github.com/tendermint/tendermint@v0.34.24/consensus/replay.go:416 +0x7ae
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc001303920, {0x2f7d008, 0xc001066820})
	github.com/tendermint/tendermint@v0.34.24/consensus/replay.go:268 +0x3d4
github.com/tendermint/tendermint/node.doHandshake({_, _}, {{{0xb, 0x0}, {0xc001185e50, 0x7}}, {0xc001185e60, 0xb}, 0x1, 0x75, ...}, ...)
	github.com/tendermint/tendermint@v0.34.24/node/node.go:329 +0x1b8
github.com/tendermint/tendermint/node.NewNode(0xc001045040, {0x2f61980, 0xc000c868c0}, 0xc000c22410, {0x2f41be0, 0xc000ebb758}, 0x1?, 0x0?, 0xc000c22740, {0x2f666a0, ...}, ...)
	github.com/tendermint/tendermint@v0.34.24/node/node.go:777 +0x597
github.com/cosmos/cosmos-sdk/server.startInProcess(_, {{0x0, 0x0, 0x0}, {0x2f8a3b8, 0xc00126e5a0}, {0x0, 0x0}, {0x2f6e2d0, 0xc001007a40}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.46.10/server/start.go:264 +0x7bb
github.com/cosmos/cosmos-sdk/server.StartCmd.func2(0xc00107d200?, {0x43b06b0?, 0x0?, 0x0?})
	github.com/cosmos/cosmos-sdk@v0.46.10/server/start.go:128 +0x193
github.com/spf13/cobra.(*Command).execute(0xc00107d200, {0x43b06b0, 0x0, 0x0})
	github.com/spf13/cobra@v1.6.1/command.go:916 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc000ef0f00)
	github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.6.1/command.go:968
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	github.com/spf13/cobra@v1.6.1/command.go:961
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0x0?, {0xc000d39608, 0x12})
	github.com/cosmos/cosmos-sdk@v0.46.10/server/cmd/execute.go:36 +0x1eb
main.main()
	./main.go:16 +0x31

```

Reply:
Starting to work on 1) in a separate PR

Reply:
Do you think we should also include this unit test, or are the e2e tests sufficient?

Reply:
I think it is also useful to have the test here as e2e is more difficult to debug if something isn't working as expected.

Reply:
Resolving this as we've addressed most steps and tracking the rest via TODOs in code